src/lsm/tree.zig

```diff
@@ -283,81 +293,211 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type) type {
+                    key_blocks.data_block_address,
+                    key_blocks.data_block_checksum,
+                    key,
+                )) {
+                    .does_not_exist => {},
+                    // Key present in the data block.
+                    .exists => |value| return .{ .exists = value },
+                    // Data block was not found in the grid cache. We cannot rule out
+                    // the existence of the key without I/O, and therefore bail out.
+                    .block_not_in_cache => return .{ .may_exist = iterator.level - 1 },
+                }
+            }
+            // Key not present in the Manifest.
+            return .does_not_exist;
+        }
+
+        fn cached_filter_block_search(
+            tree: *Tree,
+            address: u64,
+            checksum: u128,
+            fingerprint: bloom_filter.Fingerprint,
+        ) enum { does_not_exist, may_exist, block_not_in_cache } {
+            if (tree.grid.read_block_from_cache(address, checksum)) |filter_block| {
+                const filter_schema = schema.TableFilter.from(filter_block);
+                const filter_bytes = filter_schema.block_filter_const(filter_block);
+                if (!bloom_filter.may_contain(fingerprint, filter_bytes)) {
+                    return .does_not_exist;
+                } else {
+                    return .may_exist;
+                }
+            } else {
+                return .block_not_in_cache;
+            }
+        }
+
+        fn cached_data_block_search(
             tree: *Tree,
+            address: u64,
+            checksum: u128,
+            key: Key,
+        ) union(enum) {
+            exists: *const Value,
+            does_not_exist,
+            block_not_in_cache,
+        } {
+            if (tree.grid.read_block_from_cache(address, checksum)) |data_block| {
+                if (Table.data_block_search(data_block, key)) |value| {
+                    return .{ .exists = value };
+                } else {
+                    return .does_not_exist;
+                }
+            } else {
+                return .block_not_in_cache;
+            }
+        }
+
+        /// Call this function only after checking `lookup_from_memory()`.
+        pub fn lookup_from_levels_storage(tree: *Tree, parameters: struct {
             callback: fn (*LookupContext, ?*const Value) void,
             context: *LookupContext,
             snapshot: u64,
             key: Key,
-        ) void {
-            assert(tree.lookup_snapshot_max.? >= snapshot);
+            start_level: u8,
+        }) void {
             if (constants.verify) {
-                // The caller is responsible for checking the mutable table.
-                assert(tree.lookup_from_memory(snapshot, key) == null);
+                assert(tree.lookup_from_memory(
+                    parameters.snapshot,
+                    parameters.key,
+                ) == .may_exist);
```

Comment:
I think this is the same assertion problem as the previous vopr failure was.

Reply:
Indeed, this will suffer from the same issue. I should've done a search with `== .may_exist` before. 🤦🏼 

Did one now and we're good, thank you for this!